### ISSUE TITLE
[bug #18818] ATMega644P missing second USART erroneously

### ISSUE THREAD

**Issue by avrs-admin** [avrdudes/avr-libc#214]
Thu 18 Jan 2007 01:21:42 PM CET

There's a slight problem with ATMega644P and ATMega644 in the iomxx4.h header file. ATMega644 has only one USART, which is correct, but ATMega644P should have two USARTs. avr-libc seems to see both devices as the same device which they apparently are not. Even the paratetric product table at [www.atmel.com](https://savannah.nongnu.org//www.atmel.com) says that ATMega644P has only one USART, but the datasheet says otherwise.

Suggested solution: recognize ATMega164P, ATMega324P and ATMega644P as their own product family and isolate ATMega644 from it. This means adding the 'p' letter to the processors' individual headers as is already done with ATMega196 and ATMega169P. iom644.h should perhaps not be part of this family. The ATMega644 and ATMega644P should probably be seen as different devices.

This issue was migrated from https://savannah.nongnu.org/bugs/?18818

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Thu 18 Jan 2007 02:18:16 PM CET

Works as designed:

$252% cat > foo.c
#include <avr/io.h>

void foo(void)
{
UCSR1A = 42;
}
^D
$ avr-gcc -mmcu=atmega644p -c foo.c
$ avr-gcc -mmcu=atmega644 -c foo.c
foo.c: In function 'foo':
foo.c:5: error: 'UCSR1A' undeclared (first use in this function)
foo.c:5: error: (Each undeclared identifier is reported only once
foo.c:5: error: for each function it appears in.)

Are you perhaps using an obsolete version of compiler and/or
library?

**Comment by avrs-admin** [comment]
Tero Sinervo <tvjsin>
Fri 19 Jan 2007 01:50:20 AM CET

You're right.

What happened was I never bothered to check if it really works, fooled by the absence of 'p' in the individual header files, coupled with this line in iomxx4.h:

/*

- Only ATmega164/324 have a second USART.*/

I may be partly absolved by that slight glitch in the documentation, but nevertheless I shall repent my rather blatant foolishness.

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Sun 21 Jan 2007 10:11:30 PM CET

Confusing comment has been fixed in that header file.
